arch: arc: fixes the hanlding of CONFIG_EXECUTION_BENCHMARKING

vonhust opened the pull request:
when firq is enabled and there are multiple register banks,
it requires different handling

Fixes #12949 
Fixes #12947 

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: fixes the hanlding of CONFIG_EXECUTION_BENCHMARKING

when firq is enabled and there are multiple register banks,
it requires different handling

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ruuddw commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -230,47 +230,38 @@ From RIRQ:
+#endif
+/*
+ * read_timer_start_of_isr just uses several
+ * register now, no need to save all regs
```
the read_timer_isr function is not in this module, so making assumptions on its register usage is risky. Saving the context is done later anyway, maybe move the read_timer call to that point? Or just leave in the full context save - it impacts the benchmark number, but not regular performance.

ruuddw commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -230,47 +230,38 @@ From RIRQ:
+#endif
+
 #ifdef CONFIG_EXECUTION_BENCHMARKING
+#if CONFIG_RGF_NUM_BANKS !=1
```
Why switching banks here?
And more important: shouldn't the current bank (depending on entering via FIRQ or RIRQ) be saved instead of always restored to bank 1 later (line 260).

vonhust replied:
1. FIRQ's priroity is highest,  impossible to have  nest firq.  So the interruptted reg bank must be bank 0. 
2.  As some pop and push are required, we need to switch back to the interruptted bank to get the right sp.
3. switch back to bank1 means to recover the context of firq,  do not impact the following  interrupt handling

vonhust replied:
and there is a better place to call `read_timer_start_of_isr`, which is simpler,    It's at the beiging of  _isr_demux.,  but the side effect is that the perf number may be not accurate.

vonhust replied:
@ruudw, i add another simpler and cleaner commit with a result of not accurate latency number

ruuddw commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -396,39 +387,10 @@ irq_hint_handled:
 	ld_s r1, [r0, 4] /* ISR into r1 */
 #ifdef CONFIG_EXECUTION_BENCHMARKING
 	push_s r0
-	push_s blink
```
for the isr start, blink was saved as well. Not needed here? Or not needed in the earlier code for isr start?

vonhust replied:
no need to save here. Blink is safe after irq_hint_handled. Also just need to save r0 and r1

ruuddw reviewed (commented):
not fully sure if this is the right fix, please explain why the bank switching is required and why it can be done unconditionally.

Commit:
arch: arc: move the call of read_timer_start_xx to _isr_demux

It's simpler and cleaner to move read_timer_start_of_isr to

_isr_demux. But the interrupt latency may be not accurate.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ruuddw reviewed (approved):
Moving the ISR 'start' timestamp logging to the isr_demux simplifies stuff. And, it's probably not much less accurate than all the earlier additional register push/pops.
For really accurate measurements, typically HW probes are used - this should be good.

ruuddw commented:
@andrewboie @nashif : could one of you please have a look and merge? Thx.
